### DISCUSSION TITLE
Create Generated Column in MariaDB

### DISCUSSION THREAD

**Discussion by ghost** [sqlalchemy/sqlalchemy#10055] [Usage Questions]
Hello, when using [Computed Columns](https://docs.sqlalchemy.org/en/20/core/defaults.html#computed-columns-generated-always-as), the following code works without any changes on PostgreSQL, but it fails in MariaDB.

```python
import asyncio

from sqlalchemy import Computed, text
from sqlalchemy.orm import (
    Mapped,
    DeclarativeBase,
    mapped_column,
)
from sqlalchemy.ext.asyncio import create_async_engine

# works with this
# engine = create_async_engine('postgresql+psycopg://...')

# does not work with this
engine = create_async_engine('mariadb+asyncmy://...')

class Base(DeclarativeBase):
    pass

class Sqaure(Base):
    __tablename__ = 'square'
    id: Mapped[int] = mapped_column(primary_key=True)
    side: Mapped[int]
    area: Mapped[int] = mapped_column(Computed(text('4 * side')), index=True)

async def main():
    async with engine.begin() as conn:
        await conn.run_sync(Base.metadata.drop_all)
        await conn.run_sync(Base.metadata.create_all)

asyncio.run(main())
```

```
sqlalchemy.exc.ProgrammingError: (asyncmy.errors.ProgrammingError) (1064, "You have an error in your SQL syntax; check the manual that corresponds to your MariaDB server version for the right syntax to use near 'NOT NULL, \n\tPRIMARY KEY (id)\n)' at line 4")
[SQL: 
CREATE TABLE square (
        id INTEGER NOT NULL AUTO_INCREMENT, 
        side INTEGER NOT NULL, 
        area INTEGER GENERATED ALWAYS AS (4 * side) NOT NULL, 
        PRIMARY KEY (id)
)

]
(Background on this error at: https://sqlalche.me/e/20/f405)
```

What should I do so that the table gets created in MariaDB?

Making the column nullable (`area: Mapped[int | None]`) seems to fix this, but I have the following concerns

- I find it counterproductive in my case, the generated column always has a value.
- In PostgreSQL this is a "non-nullable" / required column and I'd like to not change the behavior just to accommodate MariaDB (the code has to support both dialects)

**Comment by CaselIT** [comment]
Hi,

This seems like an issue. Strange since I think this case is tested.
Do you know if mariadb requires nullable in another place or if it does not allow nullable at all on generated columns?

**Reply by ghost** [reply]
Thanks for the quick response, I was unsure so I checked the [docs](https://mariadb.com/kb/en/generated-columns/) looks like it does not allow `[NOT NULL | NULL]` in the query, is that what you were asking me? Sorry I am quite new to this.

**Reply by CaselIT** [reply]
thanks. seems like a bug indeed
